reps/2024-12-4-ray-service-incr-upgrade.md

```diff
@@ -0,0 +1,124 @@
+
+
+### Should this change be within `ray` or outside?
+Within Ray. For better code maintenance.
+
+## Stewardship
+
+### Shepherd of the Proposal
+- @kevin85421
+
+## Design and Architecture
+
+### Proposed API
+
+[target_capacity](https://github.com/ray-project/ray/blob/2ae9aa7e3b198ca3dbe5d65f8077e38d537dbe11/python/ray/serve/schema.py#L38) (int [0, 100]) - [Implemented in Ray 2.9](https://github.com/ray-project/ray/commit/86e0bc938989e28ada38faf25b75f717f5c81ed3)
+- Definition: (num_current_serve_replicas / num_expected_serve_replicas) * 100%
+- Ray Serve / RayService assumes that both the old and new RayClusters have the same capacity to handle requests. Consequently, the new RayCluster can handle up to target_capacity_new of the total requests.
+
+max_surge_percent  (int [0, 100])
+- max_surge_percent represents the maximum allowed percentage increase in capacity during a scaling operation. It limits how much the new total capacity can exceed the original target capacity. 
+    - The formula is: target_capacity_old + target_capacity_new <= (100% + max_surge_percent)
+
+After adding the above field to the Ray Serve schema, these APIs will be added to the KubeRay CR spec and can be specified by the user as follows:
+
+```sh
+type RayServiceUpgradeSpec struct {
+  // The target capacity percentage of the upgraded RayCluster.
+  // Defaults to 100% target capacity.
+  // +kubebuilder:default:=100
+  TargetCapacity *int32 `json:"targetCapacity,omitempty"`
+  // The maximum percent increase of the total capacity during a scaling operation.
+  // Defaults to 100%, i.e. a new RayCluster with equal capacity can be scaled up.
+  // +kubebuilder:default:=100
+  MaxSurgePercent *int32 `json:"maxSurgePercent,omitempty"`
+}
+```
+
+The RayService spec would then look as follows:
+```sh
+// RayServiceSpec defines the desired state of RayService
+type RayServiceSpec struct {
+  // UpgradeStrategy represents the strategy used when upgrading the RayService. Currently supports `NewCluster` and `None`
+  UpgradeStrategy *RayServiceUpgradeStrategy `json:"upgradeStrategy,omitempty"`
+  // UpgradeSpec defines the scaling policy used when upgrading the RayService with `NewCluster` strategy.
+  UpgradeSpec *RayServiceUpgradeSpec `json:"UpgradeSpec,omitempty"`
+  // Defines the applications and deployments to deploy, should be a YAML multi-line scalar string.
+  ServeConfigV2  string         `json:"serveConfigV2,omitempty"`
+  RayClusterSpec RayClusterSpec `json:"rayClusterConfig,omitempty"`
+  ...
+}
+```
+
+Example RayService manifest:
+```sh
+apiVersion: ray.io/v1
+kind: RayService
+metadata:
+  name: example-rayservice
+spec:
+  upgradeStrategy: "NewCluster"
```

Comment:
It's not too late to change this API, should we change this now before v1.3 so it'll play nicely with the new fields you're proposing here?
```
spec:
  upgradeStrategy:
    type: "NewCluster"
    targetCapacity: 50
    maxSurgePercent: 50
...
```

Reply:
At least for 1.3, we can just support
```
spec:
  upgradeStrategy:
    type: NewCluster
```

So it's easy to add new fields in v1.4 that are compatible with the `upgradeStrategy`

Reply:
a5ea9db, API change: https://github.com/ray-project/kuberay/pull/2678